devices.go

```diff
@@ -146,29 +153,61 @@ func (d *Device) GetPortByName(name string) (*Port, error) {
 	return nil, fmt.Errorf("Port %s not found in device %s", name, d.ID)
 }
 
-func (d *Device) GetNetworkType() (string, error) {
-	numOfBonds := d.NumOfBonds()
-	if (numOfBonds < 1) || (numOfBonds > 2) {
-		return "", fmt.Errorf("Wrong number of Bond ports")
+type ports map[string]*Port
+
+func (ports ports) allBonded() bool {
+	if ports == nil {
+		return false
 	}
-	bond0, err := d.GetPortByName("bond0")
-	if err != nil {
-		return "", err
+
+	if len(ports) == 0 {
+		return false
 	}
-	if numOfBonds == 2 {
-		bond1, err := d.GetPortByName("bond1")
-		if err != nil {
-			return "", err
+
+	for _, p := range ports {
+		if (p == nil) || (!p.Data.Bonded) {
+			return false
 		}
-		if bond0.NetworkType == bond1.NetworkType {
-			return bond0.NetworkType, nil
+	}
+	return true
+}
+
+func (d *Device) HasManagementIPs() bool {
+	for _, ip := range d.Network {
+		if ip.Management {
+			return true
+		}
+	}
+	return false
+}
+
+// GetNetworkType returns a composite network type identification for a device
+// based on the plan, network_type, and IP management state of the device.
+// GetNetworkType provides the same composite state rendered in the Packet
+// Portal for a given device.
+func (d *Device) GetNetworkType() string {
```

Comment:
One more thing that came to my mind. The network type is obivously most problematic in `n2.xlarge.x86`, it's got 2 bonds and 4 NICs.

It would be good to verify all the state transitions for the `n2.xlarge.x86`.

I will try to do it now, but you can too, just substitute the type in TestAccPortNetworkStateTransitions.

Reply:
Failed with:

```--- FAIL: TestAccPortNetworkStateTransitions (299.80s)
    ports_test.go:445: Failed to convert device 9c8a5d54-6383-4d49-9961-48a4222b721b from layer2-bonded to layer2-individual. New type was hybrid
FAIL
```

Last state of the device ports was

```
  "network_ports": [
    {
      "id": "a455823f-4e08-4b38-bf0f-1d12ef93f2b6",
      "type": "NetworkBondPort",
      "name": "bond0",
      "data": {
        "bonded": true
      },
      "network_type": "layer2-individual",
      "native_virtual_network": null,
      "hardware": {
        "href": "/hardware/62ec1803-fa33-4104-9d5d-883339d00b5a"
      },
      "virtual_networks": [],
      "connected_port": null,
      "href": "/ports/a455823f-4e08-4b38-bf0f-1d12ef93f2b6"
    },
    {
      "id": "f8be7677-387e-4239-8718-411487f6b4a6",
      "type": "NetworkBondPort",
      "name": "bond1",
      "data": {
        "bonded": true
      },
      "network_type": "layer2-individual",
      "native_virtual_network": null,
      "hardware": {
        "href": "/hardware/62ec1803-fa33-4104-9d5d-883339d00b5a"
      },
      "virtual_networks": [],
      "connected_port": null,
      "href": "/ports/f8be7677-387e-4239-8718-411487f6b4a6"
    },
    {
      "id": "611225a4-12fb-4058-814c-dd0671f34ea6",
      "type": "NetworkPort",
      "name": "eth0",
      "data": {
        "bonded": false,
        "mac": "e4:43:4b:6d:47:be"
      },
      "bond": {
        "id": "a455823f-4e08-4b38-bf0f-1d12ef93f2b6",
        "name": "bond0"
      },
      "native_virtual_network": null,
      "hardware": {
        "href": "/hardware/62ec1803-fa33-4104-9d5d-883339d00b5a"
      },
      "virtual_networks": [],
      "connected_port": {
        "href": "/ports/0edc734e-fab3-4462-af88-3dc574743633"
      },
      "href": "/ports/611225a4-12fb-4058-814c-dd0671f34ea6"
    },
    {
      "id": "4c6651c6-5f69-4716-a709-275960dfd1aa",
      "type": "NetworkPort",
      "name": "eth1",
      "data": {
        "bonded": false,
        "mac": "e4:43:4b:6d:47:bf"
      },
      "bond": {
        "id": "f8be7677-387e-4239-8718-411487f6b4a6",
        "name": "bond1"
      },
      "native_virtual_network": null,
      "hardware": {
        "href": "/hardware/62ec1803-fa33-4104-9d5d-883339d00b5a"
      },
      "virtual_networks": [],
      "connected_port": {
        "href": "/ports/3cbe9b56-0b12-40fe-bf2b-6d6d6ab7c21a"
      },
      "href": "/ports/4c6651c6-5f69-4716-a709-275960dfd1aa"
    },
    {
      "id": "056128c4-0164-4337-a33c-e83a652a94f8",
      "type": "NetworkPort",
      "name": "eth2",
      "data": {
        "bonded": false,
        "mac": "e4:43:4b:6d:47:c0"
      },
      "bond": {
        "id": "a455823f-4e08-4b38-bf0f-1d12ef93f2b6",
        "name": "bond0"
      },
      "native_virtual_network": null,
      "hardware": {
        "href": "/hardware/62ec1803-fa33-4104-9d5d-883339d00b5a"
      },
      "virtual_networks": [],
      "connected_port": {
        "href": "/ports/29f966ff-a493-4dd9-b425-853493eded5c"
      },
      "href": "/ports/056128c4-0164-4337-a33c-e83a652a94f8"
    },
    {
      "id": "8b1cb481-616f-4239-9b02-cf3b91181bb0",
      "type": "NetworkPort",
      "name": "eth3",
      "data": {
        "bonded": false,
        "mac": "e4:43:4b:6d:47:c1"
      },
      "bond": {
        "id": "f8be7677-387e-4239-8718-411487f6b4a6",
        "name": "bond1"
      },
      "native_virtual_network": null,
      "hardware": {
        "href": "/hardware/62ec1803-fa33-4104-9d5d-883339d00b5a"
      },
      "virtual_networks": [],
      "connected_port": {
        "href": "/ports/49ca6b1e-1b75-4262-8a13-059111089015"
      },
      "href": "/ports/8b1cb481-616f-4239-9b02-cf3b91181bb0"
    }
  ],

```

Reply:
The web logic would also return `hybrid` in this case, judging from the UI code (included in the PR description).

```json
    {
      "type": "NetworkBondPort",
      "name": "bond0",
      "data": {
        "bonded": true
      },
      "network_type": "layer2-individual",
    },
    {
      "type": "NetworkBondPort",
      "name": "bond1",
      "data": {
        "bonded": true
      },
      "network_type": "layer2-individual",
    },
```

I captured and included tests for a fully transitioned n2-xlarge to l2-individual:
https://github.com/packethost/packngo/pull/185/files#diff-9c5652fbc46b1023e4d95b133f8770efR1334-R1375

What problem does it create for this transitional state to report `hybrid`, so long as it ultimately reports `l2-individual` when it is complete, @t0mk?  Is this not like waiting for a device to go from `powering_on` to `provisioning` to `active`?

Reply:
In other words, when transitioning between `layer2-bonded` and `layer2-individual` there is a state that looks like `hybrid` (based on the definition of having bond ports bonded, but not having all data ports bonded), then perhaps `hybrid` is the correct state to report. What do you think?

Reply:
@displague Yes, there could be a time-limited wait for changed network state. I would probably keep it packngo though. I would place it in `DeviceToNetowrkType` instead of `i.client.Devices.Get(deviceID, nil)`. 

sth like 
```
func waitDeviceNetworkType(id, ntype string, c *Client) (*Device, error) {
	// 15 minutes = 180 * 5sec-retry
	for i := 0; i < 180; i++ {
		<-time.After(5 * time.Second)
		d, _, err := c.Devices.Get(id, nil)
		if err != nil {
			return nil, err
		}
		if d.GetNetworkType() == ntype {
			return d, nil
		}
	}
	return nil, fmt.Errorf("device %s is still not in state %s after timeout", id, ntype)
}
```
.. maybe with parametrized timeout and interval

Reply:
I have these changes in a new branch. https://github.com/displague/packngo/compare/network_type_parity...displague:four_nics?expand=1

I ran the tests again and l2bonded -> l2individual passed (for n3.xlarge), but the l2individual -> hybrid failed with a 422 message, which I have been seeing in the wild.

I compared notes with the portal, and it appears that the portal is also breaking all vlans when changing to hybrid.  I want to confirm this behavior before incorporating this change into my second branch:
```js
// bondPorts(responses, includePortNames, excludePortNames, bondAction = 'bond', bulkDisable = false)
        return Promise
          .all(action.device.network_ports.map(port => removeVlans(port)))
          .then(responses => bondPorts(responses, getPorts('bond', 'all'), []))
          .then(responses => convertToLayer3(responses))
          .then(responses => fetchDevice(responses))
          .then(responses => bondPorts(responses, getPorts('eth', 'odd'), getPorts('bond', 'all'), 'disbond'))
          .then(responses => callbackOnSuccess(responses, constants.NETWORK_TYPES.HYBRID))
          .catch(errors => callbackOnError(errors, constants.NETWORK_TYPES.HYBRID));
```

Reply:
I experienced this in Terraform, it seems the 422 error message may have been improved since I last encountered this:

```
Error: POST https://api.packet.net/ports/a7965acd-abf4-49de-bcad-4d395d32409f/convert/layer-3: 422 bond0 has vlans assigned 
```